File: src/utils/authValidation.ts
export type FieldErrors = Record<string, string>

const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

export function validateRegisterForm(values: {
  fullName: string
  email: string
  password: string
  confirmPassword: string
}): FieldErrors {
  const errors: FieldErrors = {}

  if (!values.fullName.trim()) {
    errors.fullName = 'Full name is required'
  }

  if (!values.email.trim()) {
    errors.email = 'Email is required'
  } else if (!EMAIL_PATTERN.test(values.email)) {
    errors.email = 'Email format is invalid'
  }

  if (!values.password) {
    errors.password = 'Password is required'
  } else if (!isValidPassword(values.password)) {
    errors.password = 'Password does not meet requirements'
  }

  if (!values.confirmPassword) {
    errors.confirmPassword = 'Confirm password is required'
  } else if (values.password !== values.confirmPassword) {
    errors.confirmPassword = 'Passwords do not match'
  }

  return errors
}

export function validateLoginForm(values: { email: string; password: string }): FieldErrors {
  const errors: FieldErrors = {}

  if (!values.email.trim()) {
    errors.email = 'Email is required'
  } else if (!EMAIL_PATTERN.test(values.email)) {
    errors.email = 'Email format is invalid'
  }

  if (!values.password) {
    errors.password = 'Password is required'
  }

  return errors
}

function isValidPassword(password: string): boolean {
  if (password.length < 8 || password.length > 20) {
    return false
  }

  return /[A-Z]/.test(password) && /\d/.test(password)
}

File: Story.API/Services/AuthValidation.cs
using System.ComponentModel.DataAnnotations;
using Story.API.DTOs;

namespace Story.API.Services;

public static class AuthValidation
{
    public static Dictionary<string, string> ValidateRegister(RegisterRequest request)
    {
        var errors = new Dictionary<string, string>();

        if (string.IsNullOrWhiteSpace(request.FullName))
        {
            errors["FullName"] = "Full name is required";
        }

        if (string.IsNullOrWhiteSpace(request.Email))
        {
            errors["Email"] = "Email is required";
        }
        else if (!new EmailAddressAttribute().IsValid(request.Email))
        {
            errors["Email"] = "Email format is invalid";
        }

        if (string.IsNullOrWhiteSpace(request.Password))
        {
            errors["Password"] = "Password is required";
        }
        else if (!IsValidPassword(request.Password))
        {
            errors["Password"] = "Password does not meet requirements";
        }

        if (string.IsNullOrWhiteSpace(request.ConfirmPassword))
        {
            errors["ConfirmPassword"] = "Confirm password is required";
        }
        else if (request.Password != request.ConfirmPassword)
        {
            errors["ConfirmPassword"] = "Passwords do not match";
        }

        return errors;
    }

    public static Dictionary<string, string> ValidateLogin(LoginRequest request)
    {
        var errors = new Dictionary<string, string>();

        if (string.IsNullOrWhiteSpace(request.Email))
        {
            errors["Email"] = "Email is required";
        }
        else if (!new EmailAddressAttribute().IsValid(request.Email))
        {
            errors["Email"] = "Email format is invalid";
        }

        if (string.IsNullOrWhiteSpace(request.Password))
        {
            errors["Password"] = "Password is required";
        }

        return errors;
    }

    public static bool IsValidPassword(string password) =>
        password.Length is >= 8 and <= 20
        && password.Any(char.IsUpper)
        && password.Any(char.IsDigit);
}
